napari/utils/events/_tests/test_typed_list.py

```diff
@@ -0,0 +1,158 @@
+import pytest
+
+from napari.utils.events.containers import (
+    EventedList,
+    NestableEventedList,
+    TypedMutableSequence,
+)
+
+
+@pytest.fixture(
+    params=[TypedMutableSequence, EventedList, NestableEventedList]
+)
+def list_type(request):
+    return request.param
+
+
+def test_type_enforcement(list_type):
+    """Test that TypedLists enforce type during mutation events."""
+    a = list_type([1, 2, 3, 4], basetype=int)
+    assert tuple(a) == (1, 2, 3, 4)
+    with pytest.raises(TypeError):
+        a.append("string")
+    with pytest.raises(TypeError):
+        a.insert(0, "string")
+    with pytest.raises(TypeError):
+        a[0] = "string"
+    with pytest.raises(TypeError):
+        a[0] = 1.23
+
+    # also on instantiation
+    with pytest.raises(TypeError):
+        _ = list_type([1, 2, '3'], basetype=int)
+
+
+def test_type_enforcement_with_slices(list_type):
+    """Test that TypedLists enforce type during mutation events."""
+    a = list_type(basetype=int)
+    a[:] = list(range(10))
+    with pytest.raises(TypeError):
+        a[4:4] = ['hi']
+    with pytest.raises(ValueError):
+        a[2:9:2] = [1, 2, 3]  # not the right length
+    with pytest.raises(TypeError):  # right length, includes bad type
+        a[2:9:2] = [1, 2, 3, 'a']
+    assert a == list(range(10)), 'List has changed!'
+
+
+def test_multitype_enforcement(list_type):
+    """Test that basetype also accepts/enforces a sequence of types."""
+    a = list_type([1, 2, 3, 4, 5.5], basetype=(int, float))
+    assert tuple(a) == (1, 2, 3, 4, 5.5)
+    with pytest.raises(TypeError):
+        a.append("string")
+    a.append(2)
+    a.append(2.4)
+
+
+def test_custom_lookup(list_type):
+    """Test that we can get objects by non-integer index using custom lookups.
+    """
+
+    class Custom:
+        def __init__(self, name='', data=()):
+            self.name = name
+            self.data = data
+
+    hi = Custom(name='hi')
+    dct = Custom(data={'some': 'data'})
+
+    a = list_type(
+        [Custom(), hi, Custom(), dct],
+        basetype=Custom,
+        lookup={str: lambda x: x.name, dict: lambda x: x.data},
+    )
+    # index with integer as usual
+    assert a[1].name == 'hi'
+    assert a.index("hi") == 1
+
+    # index with string also works
+    assert a['hi'] == hi
+
+    # index with tuple will use the `tuple` type lookup
+    assert a[{'some': 'data'}].data == {'some': 'data'}
+    assert a.index({'some': 'data'}) == 3
+    assert a[{'some': 'data'}] == dct
```

Comment:
The comment doesn't seem to match the code?

Also, does this API mean:

- a[int] is regular indexing
- a[tuple[int]] is nested indexing
- a.index(int) is custom lookup based on int/tuple, if provided?

Reply:
ah damn it... good catch.  I used a tuple at first, but yet... then I switched to a `dict` as an example (because `tuple` is going to be used for nested indexing... so it was confusing).  changed the comment to `index with a dict will use the `dict` type lookup`

> Also, does this API mean...

more generally: `int`/`slice` work as usual, `tuple` will be used for nested indexing... but the main point here is that one can change the behavior of `__getitem__` and `index` by providing a `lookup` (mapping from type to callable that returns a value) when instantiating the list, like `lookup={str: lambda x: x.name, dict: lambda x: x.data},`.  Then whenever you get an item from the list, if `type(key) in lookup`, it uses the index of the returned value instead.